### PR TITLE
Add usebreakpoint hook

### PR DESCRIPTION
**Pull request by JSReds**
See: [Flowdock discussion](https://www.flowdock.com/app/rulemotion/resin-frontend/threads/ATGc-ie1nUzCHGSbmrx8HMYIQoH)
Change-type: patch

##### Contributor checklist

- [x] I have rebuilt the README with `npm run build:docs`
- [x] I have regenerated screenshots for any affected components with `npm run generate-screenshots`
- [x] I have regenerated jest snapshots for any affected components with `npm run jest -- -u`

##### Reviewer Guidelines
- When submitting a review, please pick:
  - '*Approve*' if this change would be acceptable in the codebase (even if there are minor or cosmetic tweaks that could be improved).
  - '*Request Changes*' if this change would not be acceptable in our codebase (e.g. bugs, changes that will make development harder in future, security/performance issues, etc).
  - '*Comment*' if you don't feel you have enough information to decide either way (e.g. if you have major questions, or you don't understand the context of the change sufficiently to fully review yourself, but want to make a comment)
---

### CODE REVIEW

File: src/components/Provider/BreakpointProvider.tsx [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+export const BreakpointProvider: React.FC<BreakpointProviderProps> = ({
+	breakpoints,
+	children,
+}) => {
+	const [currentBreakpoint, setCurrentBreakpoint] = React.useState(
+		0 as Breakpoint,
+	);
+
+	const handleResize = () => {
```
Comment (sradevski) [inline src/components/Provider/BreakpointProvider.tsx]: Let's move this inside the effect

File: src/components/Provider/BreakpointProvider.tsx [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+		const clientWidth = window.innerWidth;
+		if (clientWidth < breakpoints[0]) {
```
Comment (sradevski) [inline src/components/Provider/BreakpointProvider.tsx]: I am thinking, we probably want this to be more dynamic, as the length of the `breakpoints` array can change, and the hook should be able to handle it seamlessly.

File: src/components/Provider/BreakpointProvider.tsx [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+			setCurrentBreakpoint(1);
+		} else if (clientWidth < breakpoints[2]) {
+			setCurrentBreakpoint(2);
+		} else if (clientWidth >= breakpoints[2]) {
+			setCurrentBreakpoint(3);
+		}
+	};
+
+	React.useEffect(() => {
+		const debouncedResize = debounce(handleResize, 80);
```
Comment (sradevski) [inline src/components/Provider/BreakpointProvider.tsx]: Once we move handleResize in here and make it dynamic, before any of this is called, we should validate that `breakpoints` exists and that the length is not 0 (and maybe set some maximum as well).

File: src/hooks/useBreakpoint.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import * as React from 'react';
+import { BreakpointContext } from '../contexts/BreakpointContext';
+
+export const useBreakpoint = <T extends any>(values: T[]): T => {
+	if (values.length > 4) {
```
Comment (sradevski) [inline src/hooks/useBreakpoint.ts]: Same as in the provider, we should be more flexible in accepting a range of array sizes and dynamically handle it (for example between 1 and 6)

File: src/contexts/BreakpointContext.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import * as React from 'react';
+
+export type Breakpoint = 0 | 1 | 2 | 3;
```
Comment (sradevski) [inline src/contexts/BreakpointContext.ts]: Same like the rest, we can probably make it a bit more dynamic, so maybe a plain number is fine.

File: src/components/Provider/BreakpointProvider.tsx [hunk 6]
```diff
@@ -1,44 +1,35 @@
...
-			setCurrentBreakpoint(2);
-		} else if (clientWidth >= breakpoints[2]) {
-			setCurrentBreakpoint(3);
-		}
-	};
+	const [currentBreakpoint, setCurrentBreakpoint] = React.useState<number>(0);
 
 	React.useEffect(() => {
-		const debouncedResize = debounce(handleResize, 80);
+		if (breakpoints && !!breakpoints.length) {
```
Comment (sradevski) [inline src/components/Provider/BreakpointProvider.tsx]: Let's reduce nesting by checking the opposite condition and returning early (`if (!breakpoints..) return`)

**Comment by JSReds** [comment]
@balena-ci retest

File: src/components/Provider/index.tsx [hunk 7]
```diff
@@ -22,7 +23,11 @@ const Base = styled(Grommet)`
 `;
 
 const Provider = ({ theme, ...props }: ThemedProvider) => {
-	return <Base theme={merge(defaultTheme, theme)} {...props} />;
+	return (
+		<BreakpointProvider breakpoints={defaultTheme.breakpoints}>
```
Comment (sradevski) [inline src/components/Provider/index.tsx]: We should pass the breakpoints from the merged theme here, not the default one. You would extract `merge(defaultTheme, theme)` in a variable, and use that to pass the breakpoints.

File: src/hooks/useBreakpoint.ts [hunk 8]
```diff
@@ -0,0 +1,18 @@
+import * as React from 'react';
+import { BreakpointContext } from '../contexts/BreakpointContext';
+import { breakpoints } from '../theme';
+
+export const useBreakpoint = <T extends any>(values: T[]): T => {
```
Comment (sradevski) [inline src/hooks/useBreakpoint.ts]: Let's add a comment what this is and how it should be used.

File: src/components/Provider/BreakpointProvider.tsx (hunk 6 above)
Comment (sradevski) [reply src/components/Provider/BreakpointProvider.tsx]: This doesn't seem to have been resolved, let's return early here.

**Pull request merged**